client/src/main/java/io/pravega/client/connection/impl/ConnectionPoolImpl.java

```diff
@@ -0,0 +1,207 @@
+        } else {
+            // create a new connection.
+            log.info("Creating a new connection to {}", location);
+            CompletableFuture<FlowHandler> establishedFuture = establishConnection(location);
+            connection = new Connection(location, establishedFuture);
+            prunedConnectionList.add(connection);
+        }
+        connectionMap.put(location, prunedConnectionList);
+        return connection.getFlowHandler().thenApply(flowHandler -> flowHandler.createFlow(flow, rp));
+    }
+
+    @Override
+    @Synchronized
+    public CompletableFuture<ClientConnection> getClientConnection(PravegaNodeUri location, ReplyProcessor rp) {
+        Preconditions.checkNotNull(location, "Location");
+        Preconditions.checkNotNull(rp, "ReplyProcessor");
+        Exceptions.checkNotClosed(closed.get(), this);
+
+        // create a new connection.
+        CompletableFuture<FlowHandler> handler = establishConnection(location);
+        Connection connection = new Connection(location, handler);
+        return connection.getFlowHandler().thenApply(h -> h.createConnectionWithFlowDisabled(rp));
+    }
+
+    private static boolean isUnused(Connection connection) {
+        return Futures.isSuccessful(connection.getFlowHandler()) && connection.getFlowCount() == 0;
+    }
+
+    /**
+     * Used only for testing.
+     */
+    @VisibleForTesting
+    @Synchronized
+    public void pruneUnusedConnections() {
+        for (List<Connection> connections : connectionMap.values()) {
+            for (Iterator<Connection> iterator = connections.iterator(); iterator.hasNext();) {
+                Connection connection = iterator.next();
+                if (isUnused(connection)) {
+                    connection.getFlowHandler().join().close();
+                    iterator.remove();
+                }
+            }
+        }
+    }
+
+    @VisibleForTesting
+    @Synchronized
+    public List<Connection> getActiveChannels() {
+        ArrayList<Connection> result = new ArrayList<Connection>();
+        for (List<Connection> connection : this.connectionMap.values()) {
+            result.addAll(connection);
+        }
+        return result;
+    }
+
+    /**
+     * Establish a new connection to the Pravega Node.
+     * @param location The Pravega Node Uri
+     * @return A future, which completes once the connection has been established, returning a FlowHandler that can be used to create
+     * flows on the connection.
+     */
+    private CompletableFuture<FlowHandler> establishConnection(PravegaNodeUri location) {
+        return FlowHandler.openConnection(location, clientConfig, metricNotifier, connectionFactory);
+    }
+
+    @Override
+    public void close() {
+        log.info("Shutting down connection pool");
+        if (closed.compareAndSet(false, true)) {
+            metricNotifier.close();
+            connectionFactory.close();
+            for (List<Connection> connections : connectionMap.values()) {
```

Comment:
I am not sure I understand the question. 
getClientConnection() and close() are both synchronized so they can't be called in parallel, and getClientConnection() checks if the closed flag has been set so if it is called after close it will just throw.